shared/src/commonMain/kotlin/io/customer/shared/di/KMMComponent.kt

```diff
@@ -32,16 +45,61 @@ class KMMComponent(
     private val databaseHelper: DatabaseHelper
         get() = getSingletonInstance { DatabaseHelper(databaseDriverFactory = databaseDriverFactory) }
 
+    internal val jsonAdapter: JsonAdapter
+        get() = getSingletonInstance {
+            JsonAdapterImpl(
+                logger = staticComponent.logger,
+                sdkComponent = sdkComponent,
+            )
+        }
+
     internal val trackingTaskQueryHelper: TrackingTaskQueryHelper
         get() = getSingletonInstance {
             TrackingTaskQueryHelperImpl(
                 logger = staticComponent.logger,
                 dateTimeUtil = staticComponent.dateTimeUtil,
-                jsonAdapter = staticComponent.jsonAdapter,
+                jsonAdapter = jsonAdapter,
                 databaseUtil = staticComponent.databaseUtil,
                 workspace = sdkComponent.customerIOConfig.workspace,
                 backgroundQueueConfig = sdkComponent.customerIOConfig.backgroundQueue,
                 trackingTaskDAO = databaseHelper.trackingTaskDAO,
             )
         }
+
+    internal val trackingHttpClient: TrackingHttpClient
+        get() = getSingletonInstance {
```

Comment:
Reason for this dependency being a singleton? 

I understand why our database classes are singletons as I know jetpack Room suggests using singletons (which makes me think that all SQL database implementations on Android should be singletons?) but this networking code I don't see a need to make them singletons.

Reply:
I don't see value in creating multiple instances for classes where its not really needed. Classes like `TrackingHttpClient` or `JsonAdapter` do not hold anything in memory, and probably they should never. So creating multiple instances will just be taking additional memory for their instances, and nothing else. Do you see any benefit for creating multiple instances? Or why would you want to have multiple instances for them?

Reply:
I like to keep singletons to a minimum in a codebase as they are prone to memory and state errors. 

You addressed that concern by saying that the `TrackingHttpClient` and `JsonAdapter` do not currently hold anything in memory and probably should never. If these classes were to never hold anything in memory (aka: have state) then these errors that singletons are prone to shouldn't happen. 

But I would add that (1) construction of objects is fast and cheap so constructing new instances is not a performance hit and (2) we are only human and may make a mistake in our code. What if in a future PR we introduce state to `JsonAdapter` and forget to no longer make `JsonAdapter` a singleton in the DI graph? That accident could happen and it might introduce a bug. 

So, I would rather make all classes non-singleton by default even if they don't currently hold state. Even if that class *today* doesn't have a state. I believe it would help avoid bugs in the future by accidentally being introduced.

Reply:
I wanted to add to ^^^ that I believe it's best to default to using singletons as little as possible to the point that when a singleton is used, a comment should be required in the code to explain why a singleton is needed in that object. So if the object is ever modified to no longer have the singleton feature required, we can make the object no longer a singleton.

Reply:
That can go the other way as well, e.g. what if in a future PR, we introduce some settings and expect it to be updated in all instances from one place? 😅 Just trying to say human mistakes can go either way and we can only try reducing it. 

Here, it is all about the expectations from the class. I am completely okay with moving `TrackingHttpClient` away from singleton. But I would expect `JsonAdapter` rather to behave singleton and return same results throughout the SDK. Even though construction is cheap, I don't think it is good enough reason to keep multiple objects which are not really designed or expected to hold multiple states. In fact, with multiple instances, we'll keep adding custom serializers to every instance needlessly.

For comments, I do agree, we can add more to make the expectations clear for non-util classes that are singleton.

Reply:
This comment is more of a brainstorming comment, not impacting this PR. 

From reading this conversation, I believe we are on the same page with saying - human errors can happen, as long as we have efforts to help reduce them then we should be fine. I am OK with introducing more singletons that do not hold state, such as JsonAdapter, into the code base as long as we have the code designed to help reduce the risk of introducing a bug into the code. With the way the code base is designed right now, I believe it's too easy (and therefore higher risk) to introduce a bug into the code. So, I want to propose a refactor at some point to lower this singleton/non-singleton risk. 

I have a couple of ideas off the top of my head for reducing this risk. 

### Have the DI definition exist in the dependency file

Right now, all dependencies are defined in the DI graph file. 

```kt
// KMMComponent.kt file
class KMMComponent {
  var foo: Foo 
    get() = getSingleton{}

  var bar: Bar 
    get() = getNewInstance{}
}

// Foo.kt file
class Foo {
}

// Bar.kt file 
class Bar {
}
```

If the design of `Bar` was to change and it should be a singleton or `Foo` changes and it should not be a singleton, you have to remember to go into the `KMMComponent` file and make the change. 

But what if we were to move some of the code around so it's easier to remember to change the KMMComponent code? 

```kt
// KMMComponent.kt file
class KMMComponent {
  // very simple, empty class. Only functions such as `getSingleton` and `getNewInstance` exist in this file. 
}

// Foo.kt file
val KMMComponent.foo: Foo 
  get() = getNewInstance{}

class Foo {
}

// Bar.kt file 
val KMMComponent.bar: Bar 
  get() = getSingleton{}

class Bar {
}
```

We use extensions to be able to move the KMMComponent properties (defining each dependency of the graph) into the file where the class is defined. 

I like this idea is because the class implementation (Foo/Bar) is what defines if the class should be a singleton or not. The implementation knows best if it should be a singleton or not. So, as you write code, your eyes have a greater chance (because the code exists in the same file) of seeing if the dependency is a singleton or not and you remember to edit it. 

### Use interface to define if a class should be a singleton or not 

Define an empty interface 

```kt
interface Singleton {}
```

Have each dependency inherit that interface or not

```kt
class Foo: Singleton {
}

class Bar {
}
```

Then, have logic in the DIGraph that will return a singleton or not depending on the interface:

```kt
// Today, you use `getSingleton` or `getNewInstance` functions to return a singleton or new interface
class KMMComponent {
  var foo: Foo 
    get() = getSingleton{}

  var bar: Bar 
    get() = getNewInstance{}
}

// We can change that to 1 common function that *all* properties of KMMComponent use `getInstance`:
class KMMComponent {
  var foo: Foo 
    get() = getInstance{}

  var bar: Bar 
    get() = getInstance{}
}

// The `getInstance` function has logic in it:
fun <T> getInstance(): T {
  if (dependency is Singleton) {
    return getSingleton()
  else 
    return getNewInstance
}
```

This idea I like the most because:
* It's similar to the last idea where the class implementation encapsulates if it's a singleton or not. You don't need to write code in the `Foo` file and then remember to open the `DiGraph` file and modify the code in there, too. You edit the code in just 1 file and the DIGraph just works. 

* Adding or removing an interface is easy. 

* The code should be easy and fast to read. When you read the code:
```kt
class JsonAdapter: Singleton {
}
```

You quickly understand that it's treated as a singleton if the DIGraph. So, in your code review, you can read the implementation of JsonAdapter and make suggestions if you don't believe the class should indeed be a singleton.

Reply:
This is an excellent idea. Thanks for writing this up. I'm linking this comment with DIGraph changes to make sure we don't miss this. Once we finalize on DIGraph changes, we should update graph to work like this. At quick look, I like the second idea (interface one) more. But once we are done with multiple graphs discussion, we can relook and decide.

Although this is not linked directly with multiple graph changes, but I don't want to run more than discussions on same module in parallel. So putting it just next to the multiple DI graphs discussion for now.